rust/src/parse.rs

```diff
@@ -0,0 +1,392 @@
+//! Functions for parsing byte buffers into their MCAP record contents.
+use crate::records::*;
+use std::borrow::Cow;
+use std::collections::BTreeMap;
+use std::error::Error;
+
+/// Returned from [`parse_record`] when parsing fails.
+#[derive(Debug)]
+pub enum ParseError {
+    StringEncoding(std::str::Utf8Error),
+    OpCodeNotImplemented(OpCode),
+    DataTooShort,
+}
+
+impl std::fmt::Display for ParseError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            Self::DataTooShort => write!(f, "data ended unexpectedly before end of record"),
+            Self::OpCodeNotImplemented(opcode) => write!(f, "opcode {:?} not supported", opcode),
+            Self::StringEncoding(err) => write!(f, "string field not valid utf-8: {}", err),
+        }
+    }
+}
+
+impl Error for ParseError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self {
+            ParseError::StringEncoding(err) => Some(err),
+            _ => None,
+        }
+    }
+}
+
+impl From<std::str::Utf8Error> for ParseError {
+    fn from(err: std::str::Utf8Error) -> Self {
+        ParseError::StringEncoding(err)
+    }
+}
+
+fn parse_u16(data: &[u8]) -> Result<(u16, &[u8]), ParseError> {
```

Comment:
how did you choose the pattern of returning a new slice instead of wrapping this in some kind of Reader struct and mutating state via `self.bytes` or `self.offset` or somesuch? The constant re-binding of `data` variables below feels awkward, but I haven't read a lot of Rust before.

Reply:
you're right that the re-binding of data is awkward, i could take a look at using a mutable struct here.